proof/refine/ARM/Finalise_R.thy

```diff
@@ -2329,12 +2329,162 @@ lemma replyUnlink_invs'[wp]:
   unfolding invs'_def valid_state'_def
   by wpsimp
 
+crunches schedContextDonate
+  for replies_of'[wp]: "\<lambda>s. P (replies_of' s)"
+
+lemma isHeadSome:
+ "isHead (Some h) \<longleftrightarrow> (\<exists>p. h = Head p)"
+  by (clarsimp simp: isHead_def split: reply_next.split)
+
+lemma replyPrev_list_refs_of_replies':
```

Comment:
For this one, and some the lemmas below, we probably don’t want to ask for alignment conditions in the assumption. What I mean is,  unlike `state_refs_of’`, `list_refs_of_replies’` does not require `is_aligned` and `ps_clear` for its elements. So, we don’t need to have those in the assumptions. So, in that sense, `ko_at’` is asking too much.

In fact, we already have this lemma in the form of projections. `replyPrevs_of_refs` in StateRelation.thy includes this one in one direction, and we don’t need to know the reply object specifically.